godot-ffi/src/binding/mod.rs

```diff
@@ -0,0 +1,328 @@
+/*
+ * Copyright (c) godot-rust; Bromeon and contributors.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+use crate::{
+    BuiltinLifecycleTable, BuiltinMethodTable, ClassEditorMethodTable, ClassSceneMethodTable,
+    ClassServersMethodTable, GDExtensionClassLibraryPtr, GDExtensionInterface,
+    GdextRuntimeMetadata, UnsafeOnceCell, UtilityFunctionTable,
+};
+
+// Ensure both crates are checked regardless of cfg, for the sake of development convenience.
+#[cfg_attr(not(feature = "experimental-threads"), allow(dead_code))]
+mod multi_threaded;
+#[cfg_attr(feature = "experimental-threads", allow(dead_code))]
+mod single_threaded;
+
+#[cfg(feature = "experimental-threads")]
+use multi_threaded::BindingStorage;
+#[cfg(not(feature = "experimental-threads"))]
+use single_threaded::BindingStorage;
+
+#[cfg(feature = "experimental-threads")]
+pub use multi_threaded::GdextConfig;
+#[cfg(not(feature = "experimental-threads"))]
+pub use single_threaded::GdextConfig;
+
+pub(crate) type GodotBinding = GodotBindingRaw<GdextConfig>;
+
+// Note, this is `Sync` and `Send` because all its fields are. We have avoided implementing `Sync` and `Send` for `GodotBindingRaw`
+// as that could hide issues if any of the field types are changed to no longer be sync/send, but the manual implementation for
+// `GodotBindingRaw` wouldn't detect that.
+//
+// This struct is generic over `Config`, this is only to make it possible to use two different implementations of `GdextConfig`.
+// As we cannot safely use the same `GdextConfig` in both single and multi-threaded code without requiring extra unsafe in `godot-core`.
+// This is because we'd need to implement `Sync` for `GdextConfig`, since it'd be used in the multi-threaded version. But then you could
+// share references to it between threads, which is not safe to do.
+pub(crate) struct GodotBindingRaw<Config> {
```

Comment:
It looks to me like the generic parameter could be replaced with a a `#[cfg]` on the `config` field. You have the `#[cfg]` dispatching on thread support just a few lines above, so they might as well be here.

This would also remove the need for yet another indirection, namely `type GodotBinding = GodotBindingRaw<GdextConfig>` (which btw _looks_ the same in all places, although it uses types from different modules). Thus simplifying code a bit, with 2 less names one needs to be aware of.

Reply:
If i do that then i have to `#[cfg]` out the `multi_threaded` module entirely when `experimental-threads` is disabled, as otherwise we get an error in there since `GodotBinding` doesn't implement `Sync`. Which i can do but then we do lose the development convenience of having both modules available at the same time during development.

Reply:
it could be, in that it'd make a change that makes `GodotBinding` not sync (even when `GdextConfig` is sync) become an error before hitting CI. but it's probably not a very common scenario to do that.